Add: Reorder control at the field level on the new view config UI.

jorgefilipecosta opened the pull request:
This PR adds buttons to sort directly at the field level on the new config UI.

## Screenshot
<img width="408" alt="Screenshot 2024-08-08 at 18 51 06" src="https://github.com/user-attachments/assets/349fa155-864d-40fd-9149-46c37b7dbd27">

cc: @WordPress/gutenberg-design 

## Testing Instructions
I verified the new sort buttons at the field level work as expected.

jorgefilipecosta committed:
Add: Sort control at the field level on the new view config UI.

jorgefilipecosta committed:
add sort labels

paulwilde commented:
Wouldn't these arrows make more sense sorting the order that the fields display? Seems a bit confusing that instead they seem to be duplicating the same functionality as the 'Sort by' and 'Order' fields above.

jorgefilipecosta commented:
> Wouldn't these arrows make more sense sorting the order that the fields display (left to right in the table view, for example)? Seems a bit confusing that instead they seem to be duplicating the same functionality as the 'Sort by' and 'Order' fields above.

Good point @paulwilde, I interpreted the arrows as sorting but maybe the plan was for reordering. @jameskoster could you clarify if the goal of the arrows was to sort or to reorder?

jameskoster commented:
@jorgefilipecosta yes these buttons are for re-ordering the display of fields as @paulwilde suggests. Sorry for the confusion!

Perhaps we should try using `chevronDown` and `chevronUp` to avoid overlap with the sorting buttons.

jorgefilipecosta committed:
make buttons reorder instead of sort

jorgefilipecosta commented:
> @jorgefilipecosta yes these buttons are for re-ordering the display of fields as @paulwilde suggests. Sorry for the confusion!
> 
> Perhaps we should try using `chevronDown` and `chevronUp` to avoid overlap with the sorting buttons.

Hi @jameskoster, I updated this PR to make the buttons be reorder instead of sort (only available on table layout). And I also updated the icons.

jameskoster commented:
Thanks for the update, we're getting close :)

* I noticed the tooltips still refer to sorting rather than order.
* Could we decrease the gap between the buttons to `4px`?
* Ideally the ordering/visibility buttons should only appear on hover.
* Currently the ordering buttons are disabled when the field is hidden, this causes the rows to jump around unexpectedly when visibility is toggled. To fix this I think it should be possible to set the order of all fields regardless of visibility. 
* Minor detail, the 'Featured image' label shouldn't wrap. I think setting `expanded` to `false` on the actions `HStack` should fix.

jorgefilipecosta committed:
apply feedback

jorgefilipecosta commented:
Hi @jameskoster the following items were addressed: 
- I noticed the tooltips still refer to sorting rather than order.
- Could we decrease the gap between the buttons to 4px?
- Ideally the ordering/visibility buttons should only appear on hover.
- Minor detail, the 'Featured image' label shouldn't wrap. I think setting expanded to false on the actions HStack should fix.

This issue was left out:
Currently the ordering buttons are disabled when the field is hidden, this causes the rows to jump around unexpectedly when visibility is toggled. To fix this I think it should be possible to set the order of all fields regardless of visibility.

However, only having the actions visible on hover minimizes the issue. It is not possible right now to allow setting the order of all fields regardless of visibility because the data structure that stores the field order is also the data structure that says if a field is visible, if a field is there it is visible and appears in that position.
I think we should change and have two data structures one to control the order and another one to control the visibility. Because by having a single one if we hide a field and show it again its order is changed -- the field goes to the end of the fields which is not ideal. But that is a bigger change and should be done in a separate PR if we agree with it.

jorgefilipecosta committed:
lint fixes

jameskoster committed:
Update tooltips

jameskoster commented:
I pushed a small update to the tooltips.

> I think we should change and have two data structures one to control the order and another one to control the visibility.

That sounds good, and fine to handle separately. In the mean time, instead of disabling the re-ordering buttons when a field is hidden, how about visually hiding them?

jameskoster commented on packages/dataviews/src/components/dataviews-view-config/index.tsx:
```diff
@@ -226,45 +227,174 @@ function ItemsPerPageControl() {
+								{ view.type === LAYOUT_TABLE && (
+									<>
+										<Button
+											disabled={
+												! isVisible || index < 1
+											}
+											accessibleWhenDisabled={ false }
```
Is there any reason not to use `accessibleWhenDisabled={ true }` here? Ideally the buttons are still accessible even when disabled.

jameskoster commented:
@jorgefilipecosta coming back to this point:

> Currently the ordering buttons are disabled when the field is hidden, this causes the rows to jump around unexpectedly when visibility is toggled.

Would it be possible to have separate groups for hidden / visible fields? A quick mockup:

<img width="364" alt="Screenshot 2024-08-28 at 16 53 08" src="https://github.com/user-attachments/assets/7878c0a7-fc81-44a9-8db5-b4accef7d310">

Then we can hide the up/down buttons on the hidden fields. I think that would simplify the UX quite a bit, and hopefully negate the need to modify the data structure. What do you think?

jorgefilipecosta committed:
separate hidden and visible fields

jorgefilipecosta commented:
Hi @jameskoster your feedback was applied and I separated the visible and hidden fields into two groups.

jorgefilipecosta committed:
in progress deal with combined fields

jorgefilipecosta committed:
deal with combined fields

jorgefilipecosta committed:
avoid focus loss

jorgefilipecosta committed:
lint fix

jameskoster committed:
Spacing tweak

jameskoster reviewed (approved):
Thanks Jorge. I pushed a small spacing tweak; this now looks good to go in terms of design.

oandregal commented:
@jameskoster @jorgefilipecosta what are your thoughts on fields that are no visible and not hidable?

They will never be displayed in the layout, so, as a DataViews' consumer, I wouldn't want them to be ever visible. They are still useful for providing a filter, for example. You can see this at work in the [default story for dataviews](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataviews--default) (see `type` field).

jameskoster commented:
I agree, if users cannot toggle visibility then there's no point showing the field in the view options at this stage.

oandregal commented:
> I agree, if users cannot toggle visibility then there's no point showing the field in the view options at this stage.

Prepared https://github.com/WordPress/gutenberg/pull/64999 that fixes this and a few other things.
